### PR TITLE
Make ASP.NET instrumentation lazy

### PR DESCRIPTION
**Pull request by RassK**
## Why

Towards #1437 

## What

* Makes ASP.NET loading lazy (traces + metrics)

## Tests

Existing

## Checklist

- [ ] `CHANGELOG.md` is updated.
- [ ] Documentation is updated.
- [x] New features are covered by tests.

### CODE REVIEW

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        lazyInstrumentationLoader.Add(new AspNetWebApiInitializer(InitializeOnFirstCall));
+    }
+
+    private void InitializeOnFirstCall(ILifespanManager lifespanManager)
+    {
+        if (Interlocked.Exchange(ref _firstInitialization, value: 0) != 1)
+        {
+            // InitializeOnFirstCall() was already called before
+            return;
+        }
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:44]: personal preference (maybe a little more performant?)

```suggestion
    private int _initialized;

    public AspNetInitializer(LazyInstrumentationLoader lazyInstrumentationLoader, PluginManager pluginManager)
    {
        _pluginManager = pluginManager;
        lazyInstrumentationLoader.Add(new AspNetMvcInitializer(InitializeOnFirstCall));
        lazyInstrumentationLoader.Add(new AspNetWebApiInitializer(InitializeOnFirstCall));
    }

    private void InitializeOnFirstCall(ILifespanManager lifespanManager)
    {
        if (Interlocked.Exchange(ref _initialized, value: 1) != 0)
        {
            // InitializeOnFirstCall() was already called before
            return;
        }
```

maybe `default` instead of `0` would be more readable (not sure if it will compile TBH)

let me know what you prefer

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs [hunk 2]
```diff
@@ -47,34 +50,67 @@ private class OnAssemblyLoadInitializer
         private readonly ILifespanManager _lifespanManager;
         private readonly string _requiredAssemblyName;
 
+#if NETFRAMEWORK
+        private int _firstInitialization = 1;
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:54]: can you elaborate (preferably as comments in code) why these changes are needed?

Can  `AppDomain.CurrentDomain.AssemblyLoad` can be called from multiple threads in for the same assembly on .NET Framework?

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs [hunk 3]
```diff
@@ -47,34 +50,67 @@ private class OnAssemblyLoadInitializer
...
             AppDomain.CurrentDomain.AssemblyLoad += CurrentDomain_AssemblyLoad;
+
+#if NETFRAMEWORK
+            var isRequiredAssemblyLoaded = AppDomain.CurrentDomain
+                .GetAssemblies()
+                .Any(x => GetAssemblyName(x) == _requiredAssemblyName);
+            if (isRequiredAssemblyLoaded)
+            {
+                OnRequiredAssemblyDetected();
+            }
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:72]: I understand that this code is to check if the assembly was already loaded... Maybe it should be not only for .NET Framework? I guess it could happen for .NET as well.

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs [hunk 4]
```diff
@@ -47,34 +50,67 @@ private class OnAssemblyLoadInitializer
...
-
-                AppDomain.CurrentDomain.AssemblyLoad -= CurrentDomain_AssemblyLoad;
+                OnRequiredAssemblyDetected();
+            }
+        }
+
+        private void OnRequiredAssemblyDetected()
+        {
+#if NETFRAMEWORK
+            if (Interlocked.Exchange(ref _firstInitialization, value: 0) != 1)
```
Comment (pellared) [inline src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:89]: Why this is needed?

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs (hunk 2 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:54]: There are multiple important points: 
* In .NET Framework we don't have startup hook that executes before the main content. So the startup is injected to presumably to the beginning of the main method. 
* There is a race condition between checking if the assembly is already loaded and loading event. 
* There is a race condition after the determination logic, so checking logic is doubled with the event and pre-check logic

So basically, the check is doubled not to have any slight chance where `isRequiredAssemblyLoaded = false` and assembly load event is not subscribed.

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs (hunk 1 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:44]: no preferences at all, can take yours if you like it more.

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs (hunk 3 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:72]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1605#discussion_r1019553855 first bullet.

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs (hunk 4 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:89]: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/pull/1605#discussion_r1019553855 third bullet.

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs (hunk 2 above)
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:54]: But for .NET it could also happen if one uses "another startup hook" which is executed before "OTel startup hook", right? I just think if he should not get rid of `#if NETFRAMEWORK`

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs (hunk 1 above)
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:44]: for me it is more readable as it uses "zero value" as default/initial value 😉 also it should be a lot faster (kidding)

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs (hunk 2 above)
Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:54]: Also I suggest adding the important points as comments in code 😉

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs (hunk 1 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:44]: Just in case mentioning that I got it from the `Instrument.cs`, I have no idea, why the default is 1 there 😄

File: src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs (hunk 2 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:54]: Do you think that 1:1 is good enough or you would like to adjust the wording there?

Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:54]: > But for .NET it could also happen if one uses "another startup hook" which is executed before "OTel startup hook", right? I just think if he should not get rid of #if NETFRAMEWORK

If you want to go there, I guess it gets deeper. I'm not sure if some would use different load context for eg, what happens then.

Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs:54]: > Do you think that 1:1 is good enough or you would like to adjust the wording there?

That is good enough for me 😉

**Review by pellared — APPROVED** [review]

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    private void InitializeOnFirstCall(ILifespanManager lifespanManager)
+    {
+        if (Interlocked.Exchange(ref _initialized, value: 1) != default)
+        {
+            // InitializeOnFirstCall() was already called before
+            return;
+        }
+
+        var instrumentationType = Type.GetType("OpenTelemetry.Instrumentation.AspNet.AspNetInstrumentation, OpenTelemetry.Instrumentation.AspNet");
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:46]: Because of the dependencies issue, especially with S.D.DS, it seems that is possible at the moment to get null here. The code should check for that and log as appropriate.

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+        var options = new OpenTelemetry.Instrumentation.AspNet.AspNetInstrumentationOptions();
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:48]: I'm uncertain why to get the `instrumentationType` reflection is used and here it is ok to reference the type directly. Could you please clarify?

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetMetricsInitializer.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    private void InitializeOnFirstCall(ILifespanManager lifespanManager)
+    {
+        if (Interlocked.Exchange(ref _initialized, value: 1) != default)
+        {
+            // InitializeOnFirstCall() was already called before
+            return;
+        }
+
+        var instrumentationType = Type.GetType("OpenTelemetry.Instrumentation.AspNet.AspNetMetrics, OpenTelemetry.Instrumentation.AspNet");
```
Comment (pjanotti) [inline src/OpenTelemetry.AutoInstrumentation/Loading/AspNetMetricsInitializer.cs:42]: `instrumentationType` can be null

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs (hunk 6 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:48]: The only public api in instrumentation packages is options type, everything else is internal.

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetMetricsInitializer.cs (hunk 7 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetMetricsInitializer.cs:42]: That's our shipped library, I guess it could be null only when installation corrupts? - but then we have loads of other places unprotected.

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs (hunk 5 above)
Comment (RassK) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:46]: That happens so late in the runtime, that we are probably crashed way earlier if that happens.

Comment (pellared) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:46]: Even if it throws then it should be handled here https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/abeb491aabac80b62588fdb87f7536f7520e1b71/src/OpenTelemetry.AutoInstrumentation/Loading/LazyInstrumentationLoader.cs#L69, right?

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetMetricsInitializer.cs (hunk 7 above)
Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetMetricsInitializer.cs:42]: Indirect dependencies are the scenario that this may fail. Anyway, it is the same case as the other comment.

File: src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs (hunk 6 above)
Comment (pjanotti) [reply src/OpenTelemetry.AutoInstrumentation/Loading/AspNetInitializer.cs:48]: makes sense

**Review by pjanotti — APPROVED** [review]
